### PR TITLE
[ubsan] Add -fsanitize-merge (and -fno-sanitize-merge)

### PR DESCRIPTION
**Pull request by thurstond**
'-mllvm -ubsan-unique-traps' (https://github.com/llvm/llvm-project/pull/65972) applies to all UBSan checks. This patch introduces -fsanitize-merge (defaults to on, maintaining the status quo behavior) and -fno-sanitize-merge (equivalent to '-mllvm -ubsan-unique-traps'), with the option to selectively applying non-merged handlers to a subset of UBSan checks (e.g., -fno-sanitize-merge=bool,enum).

N.B. we do not use "trap" in the argument name since https://github.com/llvm/llvm-project/pull/119302 has generalized -ubsan-unique-traps to work for non-trap modes (min-rt and regular rt).

This patch does not remove the -ubsan-unique-traps flag; that will override -f(no-)sanitize-merge.

### CODE REVIEW

File: clang/include/clang/Driver/Options.td [hunk 1]
```diff
@@ -2548,6 +2548,14 @@ def fsanitize_trap_EQ : CommaJoined<["-"], "fsanitize-trap=">, Group<f_clang_Gro
 def fno_sanitize_trap_EQ : CommaJoined<["-"], "fno-sanitize-trap=">, Group<f_clang_Group>,
                            Visibility<[ClangOption, CLOption]>,
                            HelpText<"Disable trapping for specified sanitizers">;
+def fsanitize_nonmerged_handlers_EQ
+    : CommaJoined<["-"], "fsanitize-nonmerged-handlers=">,
+      Group<f_clang_Group>,
+      HelpText<"Enable non-merged handlers for specified sanitizers">;
```
Comment (vitalybuka) [inline clang/include/clang/Driver/Options.td:2554]: Allow compiler to merged handlers

Comment (vitalybuka) [reply clang/include/clang/Driver/Options.td:2554]: There are non-_EQ version for trap and recover, we probably want them for consistency

File: clang/include/clang/Driver/Options.td [hunk 2]
```diff
@@ -2548,6 +2548,14 @@ def fsanitize_trap_EQ : CommaJoined<["-"], "fsanitize-trap=">, Group<f_clang_Gro
 def fno_sanitize_trap_EQ : CommaJoined<["-"], "fno-sanitize-trap=">, Group<f_clang_Group>,
                            Visibility<[ClangOption, CLOption]>,
                            HelpText<"Disable trapping for specified sanitizers">;
+def fsanitize_nonmerged_handlers_EQ
+    : CommaJoined<["-"], "fsanitize-nonmerged-handlers=">,
```
[the remark this answers is not in the document]
Comment (thurstond) [reply clang/include/clang/Driver/Options.td]: Done

File: clang/include/clang/Driver/Options.td (hunk 1 above)
Comment (thurstond) [reply clang/include/clang/Driver/Options.td]: Done

File: clang/lib/Driver/SanitizerArgs.cpp [hunk 3]
```diff
@@ -68,6 +68,8 @@ static const SanitizerMask TrappingSupported =
     SanitizerKind::ImplicitConversion | SanitizerKind::Nullability |
     SanitizerKind::LocalBounds | SanitizerKind::CFI |
     SanitizerKind::FloatDivideByZero | SanitizerKind::ObjCCast;
+static const SanitizerMask MergeDefault = SanitizerKind::Undefined;
+static const SanitizerMask MergeSupported = SanitizerKind::Undefined;
```
Comment (vitalybuka) [inline clang/lib/Driver/SanitizerArgs.cpp:72]: Having that this is advisory feature, maybe we don't need to be strict on support checking,
so no need to MergeSupported

Comment (thurstond) [reply clang/lib/Driver/SanitizerArgs.cpp]: Removed

File: clang/lib/Driver/SanitizerArgs.cpp [hunk 4]
```diff
@@ -696,6 +697,17 @@ SanitizerArgs::SanitizerArgs(const ToolChain &TC,
   TrappingKinds &= Kinds;
   RecoverableKinds &= ~TrappingKinds;
 
+  // Parse -f(no-)?sanitize-nonmerged-handlers flags
+  SanitizerMask AlwaysMerge; // Empty
```
Comment (vitalybuka) [inline clang/lib/Driver/SanitizerArgs.cpp:701]:
```
```
  SanitizerMask MergeKinds =
      parseSanitizeArgs(D, Args, DiagnoseErrors, MergeDefault, {},
                        {}, options::OPT_fsanitize_merge_handlers_EQ,
                        options::OPT_fno_sanitize_merge_handlers_EQ);
  remove: MergeKinds |= AlwaysMerge; // No-op
  remove: MergeKinds &= ~NeverMerge; // No-op
```

**Review by vitalybuka — APPROVED** [review]

**Pull request merged**